test/geocel/g4/VisitGeantVolumes.test.cc

```diff
@@ -0,0 +1,129 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file geocel/g4/VisitGeantVolumes.test.cc
+//---------------------------------------------------------------------------//
+#include "geocel/g4/VisitGeantVolumes.hh"
+
+#include <string>
+#include <vector>
+#include <G4LogicalVolume.hh>
+#include <G4VPhysicalVolume.hh>
+
+#include "corecel/io/Label.hh"
+
+#include "GeantGeoTestBase.hh"
+#include "celeritas_test.hh"
+
+namespace celeritas
+{
+namespace test
+{
+//---------------------------------------------------------------------------//
+struct LogicalVisitor
+{
+    std::vector<std::string> names;
+    void operator()(G4LogicalVolume const& lv)
+    {
+        names.push_back(Label::from_geant(lv.GetName()).name);
+    }
+};
+
+struct PhysicalVisitor
+{
+    std::vector<std::string> names;
+    bool operator()(G4VPhysicalVolume const& pv, int depth)
+    {
+        std::ostringstream os;
+        os << depth << ':' << Label::from_geant(pv.GetName()).name;
+        names.push_back(std::move(os).str());
+        return true;
+    }
+};
+
+class VisitGeantVolumesTest : public GeantGeoTestBase
+{
+  public:
+    using SpanStringView = Span<std::string_view const>;
+
+    SPConstGeo build_geometry() final
+    {
+        return this->build_geometry_from_basename();
+    }
+};
+
+//---------------------------------------------------------------------------//
+class FourLevelsTest : public VisitGeantVolumesTest
+{
+    std::string geometry_basename() const override { return "four-levels"; }
+};
+
+//---------------------------------------------------------------------------//
+
+TEST_F(FourLevelsTest, logical)
+{
+    LogicalVisitor visit;
+    visit_geant_volumes(visit, *this->geometry()->world());
+
+    static char const* const expected_names[]
+        = {"World", "Envelope", "Shape1", "Shape2"};
+    EXPECT_VEC_EQ(expected_names, visit.names);
+}
+
+//---------------------------------------------------------------------------//
+
+TEST_F(FourLevelsTest, physical)
+{
+    PhysicalVisitor visit;
+    visit_geant_volume_instances(visit, *this->geometry()->world());
+
+    static char const* const expected_names[] = {
```

Comment:
The reason I did breadth-first was so that we can correctly count the maximum depth... but I think that's no longer necessary. I've added another example and converted to DFS.